### ISSUE TITLE
hap codec in libavg

### ISSUE THREAD

**Issue by mrzl** [libavg/libavg#725]
Hi

I'm interested in using the HAP codec ( http://vdmx.vidvox.net/blog/hap ) within libavg.

The latest ffmpeg version 2.8 has the codec included https://github.com/FFmpeg/FFmpeg/blob/master/Changelog#L60
```
# ffmpeg -codecs | grep hap
D.VIL. hap                  Vidvox Hap decoder
```

and it works perfectly to decode/play hap encoded videos via ffplay.

Though when playing with avg_videoplayer, it's still unsupported in libavg.
```
unsupported video codec (Video: Hap, 1920x1080, 227597 kb/s).
```

Can somebody outline what needs to be done on the libavg core in order to support the HAP codec?

Thanks already,
Marcel

**Comment by uzadow** [comment]
Whether libavg supports HAP is entirely dependent on the version of ffmpeg (or libav) that libavg is linked with. So basically, all you have to do is use a current version of ffmpeg to compile our library.

**Comment by mrzl** [comment]
That's great. I just did that, while having the ffmpeg version that has hap included istalled:

```
workon venv
./configure
make -j9
make install
```

Though avg_videoplayer.py still says its an unsupported codec.

**Comment by uzadow** [comment]
Basically, libavg doesn't have knowledge of video codecs. It tells ffmpeg (by calling `avcodec_find_decoder` and `avcodec_open2` to open the video stream, and if that doesn't work, it returns with that error message. You could set the log level higher to see what's going on (`src/video/VideoDecoder.cpp:374` and `:383`).

**Comment by mrzl** [comment]
After commenting in the two mentioned lines, there's this additional output:

```
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x13fb160] Probed with size=2048 and score=100
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x13fb160] ISO: File Type Major Brand: qt  
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x13fb160] Probe buffer size limit 5000000 reached
```
After which the usual exception happens

```
Traceback (most recent call last):
  File "/home/deploy/.virtualenvs/project/bin/avg_videoplayer.py", line 142, in <module>
    app.App().run(VideoPlayer())
  File "/home/deploy/.virtualenvs/project/local/lib/python2.7/site-packages/libavg/app/app.py", line 98, in run
    self._applySettingsExtenders(kargs)
  File "/home/deploy/.virtualenvs/project/local/lib/python2.7/site-packages/libavg/app/app.py", line 213, in _applySettingsExtenders
    argvExtender.parser)
  File "/home/deploy/.virtualenvs/project/bin/avg_videoplayer.py", line 42, in onArgvParsed
    self.node.pause()
libavg.avg.Exception: /home/deploy/hap_N.mov: unsupported video codec (Video: Hap, 1920x1080, 0/1, 227597 kb/s).
```

**Comment by uzadow** [comment]
Hm, since I can't reproduce things here, I can't really help you very well. You're sure you are actually linking to that version of ffmpeg (and not, by chance, to a libav version that's installed in parallel)? Other than that, it gets harder. It's possible that ffplay contains custom code to enable hap decoding - you could try to find that.

**Comment by mrzl** [comment]
Thanks, I'll delve deeper into this soon and update here.

**Comment by uzadow** [comment]
Do you have a (preferrably small) hap sample file I can use to reproduce the error here?

**Comment by mrzl** [comment]
Yes, the smallest I could find in hap alpha: http://mrzl.net/Spiral5_Small_Spiral_only.mov

**Comment by uzadow** [comment]
Just a question before I spend time on this: What was the reason you wanted to use hap? The advertized speed advantages of hap need GPU support in the decoder, and I'm pretty sure ffmpeg doesn't implement that.

**Comment by mrzl** [comment]
I have only been using this codec on Windows and assumed porting it to ffmpeg would cover all properties of the codec. Here are some pro's and cons of the codec: http://vdmx.vidvox.net/blog/hap

I'll do a performance test on Windows and compare it with the decoder that's included in ffmpeg. I wasn't aware ffmpeg doesn't support support gpu decoding, because I dont have much experience with ffmpeg. Let me provide you with certainties regarding the playback speed improvement of that codec first--

**Comment by uzadow** [comment]
I read the blog post, and also read https://www.reddit.com/r/programming/comments/1gs8n2/introducing_daala_the_nextnextgeneration_free/ (search for hap on the page). 

What is your goal application-wise?

**Comment by mrzl** [comment]
I'm expecting really to be able to play more videos in parallel without application performance drops. Also what I've experienced when using this codec under Windows, is, that loading new videos introduces almost no delay when loaded directly from the main thread.

I've did a performance comparison and I could indeed see that the playback of hap encoded video takes up much more cpu&hdd performance. Which now officially confuses me.

Another aspect why I'm interested in this is the fact that this codec has an alpha channel version. We're lately having troubles with finding an appropriate codec to encode our videos with alpha channel. We've been using uncompressed video in the past because we couldn't find a solution. The hap codec seemed to be worth a try..
